### ISSUE TITLE
[Feature Request] PEP 602-compliant `ParamSpec(...).args` and `ParamSpec(...).kwargs` support

### ISSUE THREAD

**Issue by minmax** [beartype/beartype#404]
I was happy with `beartype_all()` while running tests, but after some updates I discovered `markupsafe` in my requirements. And something terrible happened.

prepare:
```bash
pip install beartype markupsafe
```

failure with lib:
```python
from beartype.claw import beartype_all
beartype_all()

import markupsafe
```

or with example code:
```python
from beartype.claw import beartype_all
beartype_all()

import run_me
```

run_me.py
```python
import functools
import typing as t

if t.TYPE_CHECKING:
    import typing_extensions as te

    _P = te.ParamSpec("_P")

def _simple_escaping_wrapper(func: "t.Callable[_P, str]") -> "t.Callable[_P, Markup]":
    @functools.wraps(func)
    def wrapped(self: "Markup", *args: "_P.args", **kwargs: "_P.kwargs") -> "Markup":
        return self.__class__(func(self, *args, **kwargs))

    return wrapped


class Markup(str):
    capitalize = _simple_escaping_wrapper(str.capitalize)
```

Exception:
```
beartype/_util/error/utilerrwarn.py:105: BeartypeClawDecorWarning: Method run.str.capitalize() in file "run.py" line 13:
    Traceback (most recent call last):
      File "beartype/_decor/decorcore.py", line 214, in _beartype_object_nonfatal
        return _beartype_object_fatal(obj, conf=conf, **kwargs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "beartype/_decor/decorcore.py", line 136, in _beartype_object_fatal
        beartype_nontype(obj, **kwargs)  # type: ignore[return-value]
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "beartype/_decor/_decornontype.py", line 188, in beartype_nontype
        return beartype_func(obj, **kwargs)  # type: ignore[return-value]
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "beartype/_decor/_decornontype.py", line 275, in beartype_func
        func_wrapper_code = generate_code(decor_meta)
                            ^^^^^^^^^^^^^^^^^^^^^^^^^
      File "beartype/_decor/wrap/wrapmain.py", line 122, in generate_code
        code_check_params = _code_check_args(decor_meta)
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "beartype/_decor/wrap/_wrapargs.py", line 346, in code_check_args
        reraise_exception_placeholder(
      File "beartype/_util/error/utilerrraise.py", line 137, in reraise_exception_placeholder
        raise exception.with_traceback(exception.__traceback__)
      File "beartype/_decor/wrap/_wrapargs.py", line 201, in code_check_args
        hint = sanify_hint_root_func(
               ^^^^^^^^^^^^^^^^^^^^^^
      File "beartype/_check/convert/convsanify.py", line 131, in sanify_hint_root_func
        hint = decor_meta.func_arg_name_to_hint[pith_name] = coerce_func_hint_root(
                                                             ^^^^^^^^^^^^^^^^^^^^^^
      File "beartype/_check/convert/convcoerce.py", line 141, in coerce_func_hint_root
        hint = resolve_hint(
               ^^^^^^^^^^^^^
      File "beartype/_check/forward/fwdmain.py", line 260, in resolve_hint
        func_locals = get_func_locals(
                      ^^^^^^^^^^^^^^^^
      File "beartype/_util/func/utilfuncscope.py", line 440, in get_func_locals
        func_frame_module_name = func_frame.f_globals['__name__']
                                 ~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^
    KeyError: '__name__'
```

**Comment by leycec** [comment]
> I was happy with `beartype_all()` while running tests...

:face_with_open_eyes_and_hand_over_mouth: 

**...omg.** Really? Someone's actually happy with `beartype_all()`? `beartype_all()` worked? No kidding? I'm kinda stunned. Honestly, I never expected that import hook to ever work for anyone. It was pretty much just a vanity function; I defined it because I was too proud *not* to define it. But this is the first time anybody has confirmed that hook worked for even five minutes! Admittedly, that hook stopped working as soon as you installed a new dependency. Oh, well. What you gonna do, huh?

I take what I can get around here. Today, I have learned to smile.

## In the Beginning There Was @beartype

Firstly, this:

```python
from beartype.claw import beartype_all
beartype_all()

import markupsafe
```

...is fine for me. No exceptions or warnings. Are you seeing exceptions or warnings from just that? If so, that's awful. Try upgrading to the newest @beartype `0.19.0rc0` pre-release via:

```bash
pip3 install --upgrade --pre beartype
```

Still exceptions or warnings? If so, please copy-and-paste me an exception or warning traceback. I will then be sad.

## Moving On to Other Boring Stuff That is Boring

There are a few intersecting issues here. Some of these issues are @beartype's fault. That's bad for us. But some of these issues are *not* @beartype's fault. That's bad for you. Since everybody is behaving badly, nobody has to feel bad. Bad behaviour in your `run_me.py` script includes:

```python
if t.TYPE_CHECKING:
    import typing_extensions as te

    _P = te.ParamSpec("_P")

def _simple_escaping_wrapper(func: "t.Callable[_P, str]") -> "t.Callable[_P, Markup]":
```

**...that's bad.** You've conditionally hidden the definition of the `_P` type hint behind a `t.TYPE_CHECKING` guard. This means that *only* static type-checkers like `mypy` and `pyright` will be able to parse this code. Runtime type-checkers like @beartype and `typeguard` will fail, because `t.TYPE_CHECKING` is *only* `True` for static type-checkers; that `if` conditional is `False` at runtime and thus `False` for @beartype.

To resolve this on your end, consider either:

* **Requiring `typing_extensions` as a mandatory dependency** and refactor the above logic to resemble:

    ```python
    import typing_extensions as te

    _P = te.ParamSpec("_P")

    def _simple_escaping_wrapper(func: "t.Callable[_P, str]") -> "t.Callable[_P, Markup]":
    ```

* **Requiring Python ≥ 3.10 as a mandatory dependency,** **dropping `typing_extensions` entirely,** and refactor the above logic to resemble:

    ```python
    _P = t.ParamSpec("_P")

    def _simple_escaping_wrapper(func: "t.Callable[_P, str]") -> "t.Callable[_P, Markup]":
    ```

Either's fine. You know. All up to you. The power is in your calloused hands! :fist_raised:  :zap: :fist_raised: 

## But What About That Dumb `KeyError`?

*...heh.* Yeah. We can't fool you. That's totally @beartype's fault. Let me dig deep into this tonight. It's time to roll up the sleeves and get greasy with some code that I haven't looked at in years, folks. `</gulp>`

**Comment by leycec** [comment]
**Partially resolved by 747f9a24424adc.** Technically, this resolves the horrifying `KeyError: '__name__'` exception you're seeing here. Pragmatically, this still isn't good enough. Why? Because...

@beartype doesn't actually support **PEP 602-compliant** `ParamSpec(...).args` or `ParamSpec(...).kwargs` type hints. Why? Because... nobody ever complained about that before. You are the first. Congratulations! Sucks to be the first, huh?

Let's upgrade this bug to a feature request for **PEP 602 compliance**. Since this is kinda critical, this is now my highest priority. *Why doesn't @beartype just work already?* :face_exhaling:

**Comment by minmax** [comment]
> ...is fine for me. No exceptions or warnings

ok, maybe u have different `markupsafe` version, thats what i have:

```
pip install --upgrade --pre beartype markupsafe
pip freeze
beartype==0.19.0rc0
MarkupSafe==2.1
```

and when i run

```
from beartype.claw import beartype_all
beartype_all()

import markupsafe
```

last `import` will fail with `KeyError: '__name__'`, just like i already posted in first message.

Also content of `run_me.py` was copied from `markupsafe/__init__.py` file, its not my code.

> What you gonna do, huh?

For now I've limited the list of packages to just my own (with `beartype_packages`), but maybe I should do more.


> PEP 602 compliance

It’s unreal that you took on this because of that issue. Thank you so much for your work and wonderful library.

**Comment by minmax** [comment]
>I never expected that import hook to ever work for anyone

As i, but a few months ago I just tried and it works (until yesterday).
I only needed to wrap all `typing.Protocol`s with @runtime_checkable.
Magic (and your missed hours of sleep), that's what it's about.

**Comment by leycec** [comment]
> ok, maybe u have different `markupsafe` version, thats what i have:

**Ah-ha!** That must be it. I'm on `markupsafe 2.1.5`, which... actually, that's not it. I've just manually inspected the `markupsafe.__init__` submodule for `markupsafe 2.1.5` and, indeed, it's exactly as you describe. I sigh.

Oh, well. That's life on GitHub, huh? I can't replicate this, so I'll quietly pretend I've already solved this issue for you. :smile: 

> Also content of `run_me.py` was copied from `markupsafe/__init__.py` file, its not my code.

**Ah-ha.** Now I get it. On the bright side, this means that you no longer have bugs you need to fix. Rejoice! On the dark side, this means that `markupsafe` **has bugs that it needs to fix.** In fact, `markupsafe 2.1.5` is even worse in this respect than `markupsafe 2.1`. This is what I'm seeing in the `markupsafe.__init__` submodule for `markupsafe 2.1.5`:

```python
if t.TYPE_CHECKING:
    import typing_extensions as te

    class HasHTML(te.Protocol):
        def __html__(self) -> str:
            pass

    _P = te.ParamSpec("_P")
```

That's horrible! They're not only defining the type alias `_P` but also the public protocol `HasHTML` behind a `TYPE_CHECKING` guard. I am gagging on my own tongue. This means that runtime type-checkers like @beartype and `typeguard` **cannot support current versions of `markupsafe`,** because those versions are buggy.

Basically, `markupsafe` needs to:

* **Require `typing_extensions` as a mandatory runtime dependency.** That's fine. Everybody requires `typing_extensions` as a mandatory runtime dependency. `markupsafe` devs shouldn't have any issues with that.
* **Refactor the above logic** to resemble:

```python
import typing_extensions as te

class HasHTML(te.Protocol):
    def __html__(self) -> str:
        pass

_P = te.ParamSpec("_P")
```

Would you mind submitting an upstream `markupsafe` issue about this, @minmax? You're a hero.

> For now I've limited the list of packages to just my own (with `beartype_packages`), but maybe I should do more.

...yeah. That's exactly what I'd do – *for now*. As @beartype slowly becomes more powerful and ribbed with muscle, the `beartype_all()` import hook should "eventually" become usable. However, "eventually" could be a very long time indeed. :face_holding_back_tears: 

> It’s unreal that you took on this because of that issue. Thank you so much for your work and wonderful library.

**Aww!** Thanks so much for the kind words. Unreal blood sacrifice for Python is what @beartype is all about.

> I only needed to wrap all `typing.Protocols` with `@runtime_checkable`.

**Ho, ho, ho.** The truth is finally revealed. Monkey-patching `typing.Protocol`, huh? Sounds good to me. Would you mind sharing your monkey-patch? That might be something @beartype itself should offer as a new `BeartypeConf(is_protocol_monkeypatch=True)` configuration option.

I'd think that pretty much *everybody* who wants to call `beartype_all()` would also want to enable that option. Maybe? Seems reasonable. Anyway, thanks so much again and have a great Summer.

![](https://media1.tenor.com/m/ipZaBBV85hMAAAAC/cat-hot.gif)
<sup>Jeez. Sure is boiling hot here in Canada... Even the cat is sweating buckets.</sup>

**Comment by minmax** [comment]
> Monkey-patching typing.Protocol

No, i mean all child classes, derived from `typing.Protocol`.

**Comment by minmax** [comment]
> Would you mind submitting an upstream markupsafe issue about this, @minmax?

I [will try](https://github.com/pallets/markupsafe/pull/456), but when i fix in locally, another import was failed, what a live.

`jinja2` also like to hide code in `TYPE_CHECKING` block.

```
if t.TYPE_CHECKING:
    import typing_extensions as te

 NEWLINE_SEQUENCE: "te.Literal['\\n', '\\r\\n', '\\r']" = "\n"
```

**Comment by leycec** [comment]
> because they use `typing_extensions.Self`

*...hmm.* I am squinting. `typing_extensions.Self` should still work under Python < 3.11. Isn't that the whole point of `typing_extensions`, after all? `typing` backports to older Python versions?

What do I know, though. I just eat popcorn and pet cats all day. Not exactly a glamorous lifestyle.

> this time with the library `jinja2`, `TYPE_CHECKING` block again.

*...lol.* Even `jinja2` is broke af, huh? This is why we can't have good QA. :face_exhaling: 

> Maybe it will be useful if `beartype_all` function will have a `ignore_packages: Iterable[str]` argument?

**Yes.** YES. So much yes! Excellent idea. I know I'm beating a dead horse until it's even deader here, but would you mind submitting a new @beartype feature request about that? This feature request is gettin' kinda crowded. I'm claustrophobic already.

`ignore_packages: Iterable[str]` is *exactly* what `beartype_all()` needs, though. Without that, `beartype_all()` is pretty much unusable in practice because users have no means of working around breakage in either @beartype itself or other third-party bad guys. Let's make this happen! Bear bros unite. :muscle: :bearded_person:

**Comment by leycec** [comment]
**Resolved by 0c9dc1c4f1a2daeb.** Aaaaaand... *that was shockingly brutal.* As a beneficial side-effect of being shockingly brutal, @beartype 0.19.0 not only supports PEP 612 but now also **deeply type-checks annotated variadic keyword parameters** (e.g., ``**kwargs: int``):

```python
>>> from beartype import beartype
>>> @beartype
... def awful_func(**kwargs: int) -> None: pass
>>> awful_func(this_func='is awful')
  File "<@beartype(__main__.awful_func) at 0x7fb03f0a45e0>", line 25, in awful_func
beartype.roar.BeartypeCallHintParamViolation: Function __main__.awful_func() parameter kwargs='is awful' violates type hint <class 'int'>, as str 'is awful' not instance of int.
```

Nobody ever complained about @beartype *not* doing that. Nobody annotates variadic keyword parameters, huh? Fair. That's fair. Since nobody cares, I'm playing video games all night. Of course, I'd do that anyway. Let's craft some alchemy for the kingdom, Atelier Rorona! :partying_face:
